### PR TITLE
Test Mode

### PR DESCRIPTION
**Pull request by chrisliebert**
There is some ambiguity in the current flight_mode design with direction, pin_num and value. The gpio library stores all of these in a pin object, but we Node has a Pin, a Direction and a Value.

---

This change is [Reviewable](https://reviewable.io/reviews/hgb2/psas-capstone/10)

### CODE REVIEW

**Comment by cwalexand** [comment]
Would you consider using the blank compatible gpio framework for test mode provided in the flt_mode_libs branch? However, this development isn't quite finished -- still needs to do some error handling which may take a couple days.

**Comment by chrisliebert** [comment]
@cwalexand Yes, I did consider the flt_mode_libs branch. I believe this is a compatible, or I am not understanding your question. I still plan to change the init/update methods. If you compile test_mode for this branch there are many unused warnings.

**Comment by chrisliebert** [comment]
Didn't mean to close the pull request, reopening.

**Comment by cwalexand** [comment]
Okay, but flight mode is still changing. If test mode proceeds it, this will have to change later.

**Comment by chrisliebert** [comment]
I didn't realize that you had code in the test_mode directory on /flt_mode_libs. Merged gpio.rs.

File: src/test_mode/src/gpio.rs [hunk 1]
```diff
@@ -1,8 +1,131 @@
...
+    pub fn new() -> MyPins {
+        MyPins { pins: LinkedList::new() }
+    }
+
+    pub fn add_pin(&mut self, pin_num: u64, direction: &str) {
+        let node = Node { num: pin_num, dir: direction.trim().to_string() };
+	    self.pins.push_back(node);
+    }
+
+	pub fn get_value(&mut self, pin_number: u64) -> u8 {
```
Comment (BrianBreniser) [inline src/test_mode/src/gpio.rs]: Any reason for this indentation?

**Comment by BrianBreniser** [comment]
This has been here for a week now, are there any other comments? gpio.rs has formatting issues but short of that this looks good, and even there it's just some inconsistent indentation which is an easy fix. I think this needs tests but I'm willing to merge this right now and add tests later (Since I'm helping on this side now, I'm willing to write those tests)

Any other feedback here? If not I'll merge it Thursday.

**Comment by cwalexand** [comment]
There is a compatible blank gpio template for test mode in the control_module_v1 branch that behaves identically to what is proposed for flight mode.  If you would prefer to use this, I will remove it from the control module branch.

**Comment by sanayani** [comment]
/me likey.  some thoughts:
if we want to hook up jsbsim at this point, we can replace the wrapper_init() call with one to fdm_create() & the wrapper_update() call with the one to fdm_get_version() & the wrapper_terminate() with fdm_close().  should be able to remove the wrapper_\* functions completely & just do them in rust.

will also need to update the .h file to include the extern "C" commands.  & take out the movie quote.  _lol_  s.

**Comment by chrisliebert** [comment]
There are more commits but they are no longer showing up here.

**Comment by sanayani** [comment]
fdm isn't a u64.  its a pointer to an object instantiated by jsbsim's FGFDMExec class.  it needs to be set up as an opaque struct i.e.:

constructor:
pub fn fdm_create()->*mut libc::c_void;

function:
pub fn fdm_get_version(arg: *mut libc::c_void);

deconstructor:
pub fn fdm_close(arg: *mut libc::c_void);
s.

**Pull request closed without merging**